aptos-move/aptos-aggregator/src/aggregator_extension.rs

```diff
@@ -230,52 +241,29 @@ impl Aggregator {
         // In theory, any delta will be applied to existing value. However,
         // something may go wrong, so we guard by throwing an error in
         // extension.
-        let key_bytes = id.key.0.to_vec();
-        resolver
-            .resolve_table_entry(&id.handle, &key_bytes)
-            .map_err(|_| extension_error("could not find the value of the aggregator"))?
-            .map_or(
-                Err(extension_error(
-                    "could not find the value of the aggregator",
-                )),
-                |bytes| {
-                    // Get the value from the storage to which we want to apply
-                    // the delta.
-                    let value_from_storage = deserialize(&bytes);
-
-                    // Sanity checks.
-                    debug_assert!(
-                        self.history.is_some(),
-                        "resolving aggregator with no history"
-                    );
-                    let history = self.history.as_ref().unwrap();
-
-                    // Validate history of the aggregator, ensure that there
-                    // was no violation of postcondition. We can do it by
-                    // emulating addition and subtraction.
-                    addition(value_from_storage, history.max_positive, self.limit)?;
-                    subtraction(value_from_storage, history.min_negative)?;
-
-                    // Validation succeeded, and now we can actually apply the delta.
-                    match self.state {
-                        AggregatorState::PositiveDelta => {
-                            self.value = addition(value_from_storage, self.value, self.limit)?;
-                        },
-                        AggregatorState::NegativeDelta => {
-                            self.value = subtraction(value_from_storage, self.value)?;
-                        },
-                        AggregatorState::Data => {
-                            unreachable!("history is not tracked when aggregator knows its value")
-                        },
-                    }
-
-                    // Change the state and return the new value. Also, make
-                    // sure history is no longer tracked.
-                    self.state = AggregatorState::Data;
-                    self.history = None;
-                    Ok(self.value)
-                },
-            )
+        let value_from_storage = resolver.resolve_aggregator_value(id).map_err(|e| {
+            extension_error(format!("Could not find the value of the aggregator: {}", e))
+        })?;
+
+        // Validate history and apply the delta.
+        self.validate_history(value_from_storage)?;
```

Comment:
If we validate the history, we can then unwrap the results on addition and subtraction, right?
(instead of ?)